# Principal Component Analysis

Principal component analysis is an algorithm for dimensionality reduction using feature extraction.
It combines input variables in a specific way, so that we can drop the least important variables and retain the most significant variables.

PCA can make remarkable improvements in the model performance when there are large number of variables.

read more about PCA here:
https://towardsdatascience.com/a-one-stop-shop-for-principal-component-analysis-5582fb7e0a9c

I will use logistic regression with raw data, standardized data, and PCA transformed data. Then I will compare the model performance.

In [1]:
# Importing libraries
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import seaborn as sns
%matplotlib inline

In [2]:
# loading dataset from scikit
from sklearn.datasets import load_breast_cancer

In [3]:
cancer = load_breast_cancer()

In [4]:
df = pd.DataFrame(cancer['data'],columns=cancer['feature_names'])
#(['DESCR', 'data', 'feature_names', 'target_names', 'target'])

## Data overview

In [231]:
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension           ...             worst radius  \
0                 0.07871           ...                    25.38   
1                 0.05667           ...                    24.99   
2                 0.05999           ...                    23.57   
3                 0.09744           ...                    14.91   
4                 0.05883           ...                    22.54   

   worst texture  worst perimeter  worst area  worst smoothness  \
0          17.33           184.60      2019.0            0.1622   
1          23.41           158.80      1956.0            0.1238   
2          25.53           152.50      1709.0            0.1444   
3          26.50            98.87       567.7            0.2098   
4          16.67           152.20      1575.0            0.1374   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   worst fractal dimension  
0                  0.11890  
1                  0.08902  
2                  0.08758  
3                  0.17300  
4                  0.07678  

[5 rows x 30 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
mean radius                569 non-null float64
mean texture               569 non-null float64
mean perimeter             569 non-null float64
mean area                  569 non-null float64
mean smoothness            569 non-null float64
mean compactness           569 non-null float64
mean concavity             569 non-null float64
mean concave points        569 non-null float64
mean symmetry              569 non-null float64
mean fractal dimension     569 non-null float64
radius error               569 non-null float64
texture error              569 non-null float64
perimeter error            569 non-null float64
area error                 569 non-null float64
smoothness error           569 non-null float64
compactness error          569 non-null float64
concavity error            569 non-null float64
concave points error       569 non-null float64
symmetry error             569 

In [6]:
df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension           ...             \
count     569.000000              569.000000           ...              
mean        0.181162                0.062798           ...              
std         0.027414                0.007060           ...              
min         0.106000                0.049960           ...              
25%         0.161900                0.057700           ...              
50%         0.179200                0.061540           ...              
75%         0.195700                0.066120           ...              
max         0.304000                0.097440           ...              

       worst radius  worst texture  worst perimeter   worst area  \
count    569.000000     569.000000       569.000000   569.000000   
mean      16.269190      25.677223       107.261213   880.583128   
std        4.833242       6.146258        33.602542   569.356993   
min        7.930000      12.020000        50.410000   185.200000   
25%       13.010000      21.080000        84.110000   515.300000   
50%       14.970000      25.410000        97.660000   686.500000   
75%       18.790000      29.720000       125.400000  1084.000000   
max       36.040000      49.540000       251.200000  4254.000000   

       worst smoothness  worst compactness  worst concavity  \
count        569.000000         569.000000       569.000000   
mean           0.132369           0.254265         0.272188   
std            0.022832           0.157336         0.208624   
min            0.071170           0.027290         0.000000   
25%            0.116600           0.147200         0.114500   
50%            0.131300           0.211900         0.226700   
75%            0.146000           0.339100         0.382900   
max            0.222600           1.058000         1.252000   

       worst concave points  worst symmetry  worst fractal dimension  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 30 columns]

**We can notice that there are 30 variables that can affect the likelyhood of causing cancer**


This makes it a perfect dataset for using PCA to reduce dimension and increase accuracy.

## Logistic Regression


In [8]:
X = df
y = cancer['target']

In [9]:
from sklearn.model_selection import train_test_split

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

In [11]:
from sklearn.linear_model import LogisticRegression

In [12]:
log = LogisticRegression()

In [13]:
log.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [238]:
predict_initial = log.predict(X_test)

In [239]:
from sklearn.metrics import accuracy_score, classification_report

In [240]:
a = accuracy_score(y_test,predict_initial)
a

0.935672514619883

#### We get 93.5% accuracy with the raw data. This is not bad but let's use standard scaled data

## Standard Scaled Data 

In [ ]:
from sklearn.preprocessing import StandardScaler

In [255]:
scaler = StandardScaler()
scaled_df = scaler.fit(df)
scaled_data = scaler.transform(df)

In [256]:
df_scaled = pd.DataFrame(scaled_data,columns=cancer['feature_names'])
df_scaled.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0     1.097064     -2.073335        1.269934   0.984375         1.568466   
1     1.829821     -0.353632        1.685955   1.908708        -0.826962   
2     1.579888      0.456187        1.566503   1.558884         0.942210   
3    -0.768909      0.253732       -0.592687  -0.764464         3.283553   
4     1.750297     -1.151816        1.776573   1.826229         0.280372   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0          3.283515        2.652874             2.532475       2.217515   
1         -0.487072       -0.023846             0.548144       0.001392   
2          1.052926        1.363478             2.037231       0.939685   
3          3.402909        1.915897             1.451707       2.867383   
4          0.539340        1.371011             1.428493      -0.009560   

   mean fractal dimension           ...             worst radius  \
0                2.255747           ...                 1.886690   
1               -0.868652           ...                 1.805927   
2               -0.398008           ...                 1.511870   
3                4.910919           ...                -0.281464   
4               -0.562450           ...                 1.298575   

   worst texture  worst perimeter  worst area  worst smoothness  \
0      -1.359293         2.303601    2.001237          1.307686   
1      -0.369203         1.535126    1.890489         -0.375612   
2      -0.023974         1.347475    1.456285          0.527407   
3       0.133984        -0.249939   -0.550021          3.394275   
4      -1.466770         1.338539    1.220724          0.220556   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0           2.616665         2.109526              2.296076        2.750622   
1          -0.430444        -0.146749              1.087084       -0.243890   
2           1.082932         0.854974              1.955000        1.152255   
3           3.893397         1.989588              2.175786        6.046041   
4          -0.313395         0.613179              0.729259       -0.868353   

   worst fractal dimension  
0                 1.937015  
1                 0.281190  
2                 0.201391  
3                 4.935010  
4                -0.397100  

[5 rows x 30 columns]

In [257]:
X_train, X_test, y_train, y_test = train_test_split(scaled_data, y, test_size=0.30)
log_std = log.fit(X_train,y_train)
predict_2 = log_std.predict(X_test)

In [258]:
b = accuracy_score(y_test,predict_2)
b

0.9649122807017544

#### We notice that the accuracy increased slightly to 96% using stadardized data set . Not bad!

## PCA Trasformed Data

In [259]:
from sklearn.decomposition import PCA

In [260]:
pca = PCA()
pca.fit(df_scaled)


PCA(copy=True, iterated_power='auto', n_components=None, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)

In [261]:
X_pca = pca.transform(df_scaled)

In [264]:
X_train, X_test, y_train, y_test = train_test_split(X_pca, y, test_size=0.30)
log.fit(X_train,y_train)
predict_3 = log.predict(X_test)

In [265]:
c = accuracy_score(y_test,predict_3)
c

0.9883040935672515

#### The Accuracy increased to a whopping 98.8%. This is remarkable improvement.

## Model comparision 

In [293]:
accuracy = []
accuracy.append(a)
accuracy.append(b)
accuracy.append(c)

methods = ['Direct data','Standardized data','Stardardized data with PCA']

In [299]:
acc_df = pd.DataFrame(accuracy,methods)
acc_df

0
Direct data                 0.935673
Standardized data           0.964912
Stardardized data with PCA  0.988304